### DISCUSSION TITLE
Interrupted Signing for ZealiD

### DISCUSSION THREAD

**Discussion by KevinKewangYu** [MatthiasValvekens/pyHanko#161] [Q&A] [answered]
In my application, I need to implement interrupted signing for ZealiD. 
I uploaded prep_digest.document_digest to ZealiD server to create the signature. The following codes show how I created prep_digest.document_digest.

prep_digest, signed_attrs, psi, output = await prep_doc(input_pdf, cert_path, timestamper_url, timestamper_timeout)
doc_digest = str(prep_digest.document_digest)

The signature I got is (look like in PEM format),
"bZRN6CAE85xgKJ0LkHqqbhkfgOvW5Pi+7Q669EJvsUaliUmzTQrgKfIDn/n6UW6m\r\nWxI3HBaxjYBNQDAWQ00ATFml6LeACXv1xhEhDYjURkdpQPcC/9z2LthB2+MEJZM1\r\n92LtrGt65KmTbcZ2YSxm/Ank1/hRDRl2nJ2sePzfmOfw65jQ97T2s72eI50t9aiW\r\nGiu/84O51J/uQ1UDscDIT3SwHaWIJAygb9dV6BZ/i2CY3ziHlIJRSRzDYinvCBb4\r\nIberDYzsimUOs9SAQupjGRmD/qdU9foSz+WFtb1vtGnqKrYW7VKPNalclsVKeMyc\r\nHQe5zENUqnAyWiRdSsBZMSJ85JH3UsEs6Cm0pxcg6frJCoDKRdF4gjaAC460/xH9\r\nXSgT/dvATzgPr8Wck1tuWo4l6JpO4LfaF5YO3LFeRsTTylytZ09JVCvoKWoqweRb\r\n6gnpQtH+A2EE90GXBUUGE6zLh0YzDPYS/MyPzexGe33PUjmBomQ/UpfsBan5Rsf/\r\nAGq+CxLZ38GOwJE35gqHXeyQsbZcSm16NxuXyC8gkBGNTzdUg8C2hvbXGCpdpdfO\r\nxtdDmmbyDZUqC4ufmHPKC7yfn2njyHbXQsKX26eaGF0vIaFOdcY7S+dyJ0t0xPyR\r\n8vTA/KCdLVIOQvalfPtilNxBp6RC9iWPxyDpDqST2Zo="

But when I set it to external_signer to generate the final digitally signed PDF file, the signature in PDF file is invalid. The following codes show what I did.

ext_signer = instantiate_external_signer(signing_certificate, sig_value)
signed_attrs = await ext_signer.signed_attrs(
    prep_digest.document_digest, 'sha256', use_pades=True
)
signed_attrs.load(signed_a)
sig_cms = await ext_signer.async_sign_prescribed_attributes(
    'sha256', signed_attrs=signed_attrs,
     timestamper=timestamps.HTTPTimeStamper(timestamper_url, https=False, timeout=timestamper_timeout)
)
vc = ValidationContext(allow_fetching=True, extra_trust_roots=[signing_certificate])
await PdfTBSDocument.async_finish_signing(
    output_handle, prepared_digest=prep_digest,
    signature_cms=sig_cms,
    post_sign_instr=psi,
    validation_context=vc
)

I also attached the signed PDF file.
[8a8332faab654f4a9eac73c07461c7de.pdf](https://github.com/MatthiasValvekens/pyHanko/files/9674145/8a8332faab654f4a9eac73c07461c7de.pdf)

I found that the interrupted signing demo provided by pyHanko used signed_attrs.dump() to simulate the remote signing.

sig_value = await sim_sign_remote(signed_attrs.dump())

async def sim_sign_remote(data_tbs: bytes):
    priv_key = serialization.load_pem_private_key(
        data,
        password=None
    )
    padding = PKCS1v15()
    hash_algo = hashes.SHA256()
    return priv_key.sign(data_tbs, padding, hash_algo)

Anyting wrong from my steps?

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @KevinKewangYu


That signature value you showed is base64-encoded. From a cursory inspection of the file you attached, it appears as if you fed pyHanko the base64-encoded version of the signature instead of the "actual" signature value.

You probably need to change this line:

```python
ext_signer = instantiate_external_signer(signing_certificate, sig_value)
```

to

```python
ext_signer = instantiate_external_signer(signing_certificate, base64.b64decode(sig_value))
```

for it to work (after importing `base64`, obviously).

Hope that helps.

**Reply by KevinKewangYu** [reply]
Hi @MatthiasValvekens , thanks for your quick reply. I have tried to b64decode it directly but cannot work. The signature I got is not a simple base64 string. It looks like in PEM format. Is there a method to extract the sig_value from such a signature of PEM format？

**Reply by peteris-zealid** [reply]
```
base64.b64decode(sig_value.replace("\r\n", ""))
```

**Reply by peteris-zealid** [reply]
@MatthiasValvekens how did you figure this out? Because I checked the cms for both of these strings and they were not there.
```
import base64
base64.b64decode("bZRN6CAE85xgKJ0L").hex()  # '6d944de82004f39c60289d0b'
bytes("bZRN6CAE85xgKJ0L", "utf8").hex()  # '625a524e36434145383578674b4a304c'
```

**Reply by MatthiasValvekens** [reply]
@peteris-zealid I simply ran the whole CMS payload through an ASN.1 decoder and noticed that the byte length of the signature in the CMS was off by a weird number of bytes, and the contents looked vaguely like base64. I didn't really try to match it up to the signature value in the OP (which could be from a different run for all I know).

I could confirm that the value in the OP decodes to a 512-byte string when passed through b64decode, though (Python's base64 module ignores spurious whitespace).

> Is there a method to extract the sig_value from such a signature of PEM format？

To the best of my knowledge, there's no such thing as PEM for raw signature values :). And even if there were, it'd still be base64 with linebreaks, so `b64decode()` should do the trick.

(Can't offer much more than generalities right now because I'm at home sick, if it's not resolved by the time I recover I'll have a better look :) )

**Reply by peteris-zealid** [reply]
If you are sick I suggest you stay away from the computer and get some rest. I wish for you to get well soon.

There is an ancient Chinese saying that music is medicine. Give it a try.

**Reply by KevinKewangYu** [reply]
Hi @MatthiasValvekens @peteris-zealid , please allow me to update the latest signed PDF file, which was created by following latest steps according to your comments,
1. hash_sha256 = hashlib.sha256(signed_attrs.dump())
2. I base64-encoded hash_sha256 and then uploaded it to ZealiD to get the signature
3. After I got the signature, base64-decoded it,
   sig_value_decoded = base64.b64decode(sig_value.replace("\r\n", ""))
4. I set sig_value_decoded to pyHanko external_signer,
   ext_signer = instantiate_external_signer(signing_certificate, sig_value_decoded)

[4e74eafb15674ff787fcbb2b36be7f18.pdf](https://github.com/MatthiasValvekens/pyHanko/files/9696891/4e74eafb15674ff787fcbb2b36be7f18.pdf)

As you can see from the latest signed PDF file, the signature is still invalid. Look forward to your update. Thanks!

**Reply by MatthiasValvekens** [reply]
(back among the living)

Hmmm, that process looks sensible at first sight, and I couldn't find an obvious smoking gun in the file. Some things you might still want to double-check:

 - Are the `signed_attrs` universally tagged? They usually are by default, but if you're retrieving them in some roundabout way they might not be. To make sure, try `signed_attrs.untag().dump()`.
 - Double check whether the API you're talking to expects pre-hashed data or not (although I suspect @peteris-zealid would've told you already if you were getting that one wrong)
 - Are you sure the certificate you declare as the signer's certificate actually matches the credentials you're using to sign?

I'm getting the feeling that you'll need some ZealiD-specific support, though... I can't really help you with that, but @peteris-zealid might be able to set you up (some engineers at ZealiD are quite familiar with pyHanko ;) ).

**Reply by KevinKewangYu** [reply]
> Hi @MatthiasValvekens @peteris-zealid , please allow me to update the latest signed PDF file, which was created by following latest steps according to your comments,
> 
> 1. hash_sha256 = hashlib.sha256(signed_attrs.dump())
> 2. I base64-encoded hash_sha256 and then uploaded it to ZealiD to get the signature
> 3. After I got the signature, base64-decoded it,
>    sig_value_decoded = base64.b64decode(sig_value.replace("\r\n", ""))
> 4. I set sig_value_decoded to pyHanko external_signer,
>    ext_signer = instantiate_external_signer(signing_certificate, sig_value_decoded)
> 
> [4e74eafb15674ff787fcbb2b36be7f18.pdf](https://github.com/MatthiasValvekens/pyHanko/files/9696891/4e74eafb15674ff787fcbb2b36be7f18.pdf)
> 
> As you can see from the latest signed PDF file, the signature is still invalid. Look forward to your update. Thanks!

Hi @MatthiasValvekens @peteris-zealid This approach can work now. The issue of invalid signature was caused by passing corrupted hash data between modules (JAVA&Python). Thanks a lot for your help. Without your help, we can't complete the interrupted signing process. As for now, this is just tech POC and far from product. So regarding the topic of financial contributions, I think we can discuss further from business perspective.

**Reply by MatthiasValvekens** [reply]
Glad to hear that you found the issue!

**Comment by peteris-zealid** [comment]
This is the first precedent of someone asking about ZealiD specific things on a public forum.
That caused some organizational/communication confusion on our side.

Most likely the error is caused by providing the hash of the prepared pdf not the hash of the signed attributes (the ASN1 CMS sub-structure) as Mathias already mentioned.

That being said what you probably want to do is call the ZealiD API differently and NOT use the CSC standard.
When calling `/oauth2/authorize` and `/signatures/signHash` in the json data add a key `hash_type` with value `PDF_HASH`. The response will then have a key called `cms` with an array of base64 encoded ASN1 structures.
The code should look something like this:
```
pdf_signer = signers.PdfSigner(
    signers.PdfSignatureMetadata(),
    signer=signers.ExternalSigner(None, None, bytes(32)),
)
prep_digest, tbs_document, output = pdf_signer.digest_doc_for_signing(pdf_out)
cms_array = call_zealid(sha256(prep_digest).digest())
PdfTBSDocument.finish_signing(
    output,
    prepared_digest=prep_digest,
    signature_cms=cms_array[0],
)
with open("result.pdf", "wb") as f:
    f.write(output.read())
```
Note that in this case there is no need to know the certificates ahead of time.

If following the CSC spec is a must have for you then I recommend looking at the CscSigner in pyHanko. It should take care of all the generic stuff. We at ZealiD did not adapt this signer to our service because we felt it was not a priority (partially due to the assumption that everyone else in the industry writes Java) but also because nobody asked for it explicitly. Asking ZealiD to write this is an option. We have enough knowledge to do it.

This way of communicating is really wooden, but I do not see another way. I cannot work on whatever I feel like because from the company perspective there is an unclear line between supporting clients and doing their work for them. In other words to spend more than a day to help you I need orders from above.

Either way pyHanko will do the heavy lifting for you. As such there is a moral obligation to support the library financially if it is used for profit.

Post Mortem: signing pdfs requires very specific knowledge that even a well known pdf software company might not have. Having worked with pdfs a lot they assume they know how to embed a signature in a pdf (how difficult can it be right?). I was informed that telling clients that they do not know how to embed an element into a pdf is not good for sales. This is how we end up with sub-optimal technical solutions.

**Reply by MatthiasValvekens** [reply]
Good summary; a couple side notes from me:

> Most likely the error is caused by providing the hash of the prepared pdf not the hash of the signed attributes (the ASN1 CMS sub-structure) as Mathias already mentioned.

Not sure this is it tbh, since OP's last comment pretty explicitly has them hashing the dump of the signed attrs. That should work. But if ZealiD has a "batteries included" API that just returns a complete CMS object from the document hash only, that might be easier in some cases (see below for a caveat).

> Note that in this case there is no need to know the certificates ahead of time.

Not unimportant: omitting `signing_cert` requires the latest release (0.14.0, see #141). Not supplying the certificates ahead of time also means that you'll have to trigger a bunch of PAdES-related things manually, which might be an issue for the OP.

> If following the CSC spec is a must have for you then I recommend looking at the CscSigner in pyHanko.

If you end up going this route, @KevinKewangYu, then I'd love some feedback regarding your experience with it, because CSC is tricky to implement generically (and I've only used pyHanko's implementation against one "real" CSC service so far).

> This is the first precedent of someone asking about ZealiD specific things on a public forum.
That caused some organizational/communication confusion on our side. [...] This way of communicating is really wooden, but I do not see another way. I cannot work on whatever I feel like because from the company perspective there is an unclear line between supporting clients and doing their work for them. In other words to spend more than a day to help you I need orders from above.

Navigating this is indeed difficult for many companies :). The line between having community support that works and avoiding spending too much engineering time on "free" support isn't always clear... Also, you don't always know where your community discussions happen: could be on StackOverflow, discussion pages of upstream projects (like in this case), mailing lists, Reddit, bulletin boards... Best of luck to ZealiD in figuring out a good way to handle that :).

Anyway, for me personally, I'm OK with this kind of discussion happening over here as long as it doesn't turn into a support forum for ZealiD-specific issues. So far I don't see any danger of that happening, so carry on 👍 . 

> Either way pyHanko will do the heavy lifting for you. As such there is a moral obligation to support the library financially if it is used for profit.

While I certainly appreciate financial contributions, I wouldn't have used the term "moral obligation" myself :). That said, there are certain aspects of maintaining a project like this that cost Actual Money™ (mostly standards-related stuff & keeping up with current tech), and obviously the time commitment is also significant, so structural contributions would for sure be helpful.

> Post Mortem: signing pdfs requires very specific knowledge that even a well known pdf software company might not have. 

Heh, the set of people that understand PDF _and_ digital signing is indeed pretty small (granted: the design philosophy behind PDF signatures isn't exactly straightforward). Even in companies that "should" know better, institutional knowledge can be quite fragmented and spread around inefficiently, esp. in product support conversations. It's not just the PDF industry, by the way. I've had some very similar interactions with well-known PKI vendors as well...

We can swap war stories next time I'm in Vilnius ;).
